cmd/goplugin-helm/resource/release.go

```diff
@@ -0,0 +1,177 @@
+	}
+	var releaseList ReleaseList
+	b := []byte(s)
+	err = json.Unmarshal(b, &releaseList)
+	if err != nil {
+		panic(fmt.Errorf("unable to unmarshal, err is %v and output is %s", err, b))
+	}
+	log.Debug("Read Releases", "externalID", externalID, "output", s)
+	for _, v := range releaseList.Releases {
+		if extID(v) == externalID {
+			return &v, nil
+		}
+	}
+	return nil, serviceapi.NotFound("release", externalID)
+}
+
+// Update an existing release
+func (h *ReleaseHandler) Update(externalID string, desiredState *Release) (*Release, error) {
+	hclog.Default().Debug("Updating Instance", "externalID", externalID, "desiredState", desiredState)
+	externalID2 := extID(*desiredState)
+	if externalID != externalID2 {
+		return nil, fmt.Errorf("ExternalID mismatch between %v and %v", externalID, externalID2)
+	}
+	return h.upsertRelease(externalID, desiredState)
+}
+
+// Delete an existing release
+func (*ReleaseHandler) Delete(externalID string) error {
+	log := hclog.Default()
+	log.Debug("Deleting Release:", "externalID", externalID)
+
+	args := []string{
+		"delete",
+		externalID,
+		"--purge",
+		// "--wait", // https://github.com/helm/helm/issues/2378 --wait is not accepted on delete, yet
+	}
+	if log.IsDebug() {
+		args = append(args, "--debug")
+	}
+	_, err := runCmd("", "helm", args...)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *ReleaseHandler) upsertRelease(externalID string, desiredState *Release) (*Release, error) {
+	log := hclog.Default()
+	namespace := ns(desiredState)
+
+	// HACK: read the release and delete it if status is failed
+	// as a workaround because --force does not replace a failed release
+	// https://github.com/helm/helm/issues/3353#issuecomment-460112405
+	// but we cannot wait for the deletion to complete, so there is a good chance that
+	// dependent resources such as persistentvolumeclaims will not be deleted in time
+	r, _ := h.Read(externalID) //TODO check the err here is not a NotFound before proceeding
+	if r != nil && r.Status != nil && *r.Status == "FAILED" {
+		log.Debug("Deleing release %v because the status is FAILED", externalID)
+		err := h.Delete(externalID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	args := []string{
+		"upgrade",
+		"--install",
+		"--force",
+		desiredState.Name,
+		desiredState.Chart,
+		"--namespace",
+		namespace,
+		"--wait",
+	}
+	if log.IsDebug() {
+		args = append(args, "--debug")
+	}
+	if desiredState.Timeout != nil {
+		args = append(args, "--timeout", strconv.Itoa(*desiredState.Timeout))
+	}
+	if desiredState.Overrides != nil {
+		for k, v := range *desiredState.Overrides {
+			args = append(args, "--set")
+			kv := fmt.Sprintf("%v=%v", k, v)
+			args = append(args, kv)
+		}
+	}
+	_, err := runCmd("", "helm", args...)
+	log.Debug("completed helm upgrade", "err", err)
+	if err != nil {
+		return nil, err
+	}
+	return h.Read(extID(*desiredState))
+}
+
+func runCmd(dir string, name string, arg ...string) ([]byte, error) {
```

Comment:
Nit, but `arg` should probably be `args` here too.

Reply:
yep.